File: 8.回溯/20.N皇后/51.n-皇后.cpp
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N 皇后
 */
#include "tools.h"
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<string> pack(n, string(n, '.'));
        auto check = [&](int row, int col) -> bool
        {
            for (int i = 0; i < n; i ++ )
            {
                if (pack[i][col] == 'Q') return 0;
            }
            for (int i = row, j = col; i >= 0 && j >= 0; i --, j --)
            {
                if (pack[i][j] == 'Q') return 0;
            }
            for (int i = row, j = col; i >= 0 && j < n; i --, j ++ )
            {
                if (pack[i][j] == 'Q') return 0;
            }
            return 1;
        };
        auto dfs = [&](auto&& self, int row) -> void
        {
            if (row == n)
            {
                res.emplace_back(pack);
                return;
            }
            for (int col = 0; col < n; col ++ )
            {
                if (check(row, col))
                {
                    pack[row][col] = 'Q';
                    self(self, row + 1);
                    pack[row][col] = '.';
                }
            }
        };
        dfs(dfs, 0);
        return res;
    }
};
// @lc code=end

File: 8.回溯/20.N皇后/51.n-皇后.py
#
# @lc app=leetcode.cn id=51 lang=python3
#
# [51] N 皇后
#
from mytools import *
# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        cols = set()
        diag1 = set()
        diag2 = set()
        res = []
        pack = ['.' * n for _ in range(n)]
        def dfs(row, res: List, pack: List):
            if row == n:
                res.append(pack[:])
                return
            for col in range(n):
                if col in cols or (row - col) in diag1 or (row + col) in diag2: continue
                pack[row] = pack[row][:col] + 'Q' + pack[row][col + 1:]
                cols.add(col)
                diag1.add(row - col)
                diag2.add(row + col)
                row += 1
                dfs(row, res, pack)
                row -= 1
                pack[row] = pack[row][:col] + '.' + pack[row][col + 1:]
                cols.remove(col)
                diag1.remove(row - col)
                diag2.remove(row + col)
        dfs(0, res, pack)
        return res
# @lc code=end

